better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+
+
+Documentation is also essential for local reasoning.  We build atop
+libraries and a programming language, which run on an operating
+system, which runs on computer hardware, which ultimately depends on
+the laws of physics.  So what keeps us from recursing, non-locally,
+down to the limits of known physics when we think about how a program
+works?
+
+#### Building a Tower of Abstraction
+
+The answer is documentation.  We can use libraries and our programming
+language without digging into their implementations because they are
+documented. Compiler engineers can do their jobs because the hardware
+manufacturers document the architecture and instruction sets. Hardware
+designers succeed because physicists document the laws of physics.
+That's the tower we're building.  Each layer of documentation defines
+an abstraction boundary that eliminates the need to research an
+implementation when reasoning about code. Without it, quality software
+development at scale becomes infeasible.
+
+The catch is that the code you are writing is almost never at the top
+of the tower.  Someone else, or future-you, is going to have to build
+on the code you're writing. There's no difference between public and
+private stuff where documentation is concerned.  Every component is
+API to somebody, so when we say “API” we mean any interface boundary
+in the codebase, however internal.
+
+#### Contract Checking Features
+
+You may have heard that some languages have features to support “Design
+by Contract.”  In general, that means you can write *parts* of your
+contracts as code, and get some checking at runtime to make sure these
+contracts are upheld. [^compile_time_checks]
+
+[^compile_time_checks]: Languages exist that check contracts at
+compile time, such as Dafny and Lean 4, but we do not think they are
+practical tools for programming at scale.
+
+```swift
+struct MyArray<T> {
+  ...
+
+  // Returns the `i`th element.
+  @requires(i >= 0 && i < self.count)
+  fun getNth(i: Integer): T
+
+  ...
+}
+```
+
+The idea started with Bertrand Meyer's own Eiffel language, and was
+picked up by many others, including D, Scala, Kotlin, and
+Clojure. Other languages, like Rust and Python, have contract
+libraries that provide a near-native contract programming experience.
+
+If you use one of these languages, fantastic; *absolutely do* leverage
+those features and libraries.  That can reduce the amount of pure
+documentation you need to write and add automated contract checking at
+runtime. But there are caveats:
+
+1. **Contracts are fundamentally documentation**, even if they're
+   expressed in code, so they must appear in the API descriptions
+   consumed by client programmers. If you're using contract checking
+   with automated documentation extraction tools make sure they expose
+   the contract code along with the API.  If not, you need to repeat
+   the complete contract in documentation.
+
+2. Some contracts are better expressed in English than as code,
```

Comment:
```suggestion
2. Some contracts are better expressed in natural language than as code,
```

I'm curious what you think of this - though it's massively dominant in the industry, English is not the only language in which code and documentation are written.

Reply:
I agree with the spirit of the suggestion, but think I'd prefer this:

```suggestion
2. Some contracts are better expressed in human language than as code,
```